frontend/src/metabase/visualizations/echarts/cartesian/waterfall/option/index.ts

```diff
@@ -1,49 +1,274 @@
+
+        const verticalAlignOffset =
+          CHART_STYLE.seriesLabels.size / 2 + CHART_STYLE.seriesLabels.offset;
+
+        const hasBottomSpace =
+          rect.y +
+            CHART_STYLE.seriesLabels.size +
+            CHART_STYLE.seriesLabels.offset <
+          chartMeasurements.bounds.bottom;
+
+        const barHeight = rect.height;
+        const endSign = getNumberOr(end, 0) < 0 ? 1 : -1;
+        let labelOffset = (endSign * barHeight) / 2;
+        labelOffset +=
+          isIncrease || !hasBottomSpace
+            ? -verticalAlignOffset
+            : verticalAlignOffset;
+
+        return {
+          hideOverlap: settings["graph.label_value_frequency"] === "fit",
+          dy: labelOffset,
+        };
+      },
+      encode: {
+        y: WATERFALL_END_KEY,
+        x: X_AXIS_DATA_KEY,
+      },
+      label: buildLabelOption(WATERFALL_VALUE_KEY),
+      animationDuration: 0,
+    },
+  ];
+
+  if (settings["waterfall.show_total"]) {
+    series.push({
+      id: WATERFALL_TOTAL_KEY,
+      type: "bar",
+      barWidth,
+      zlevel: CHART_STYLE.series.zIndex + 10,
+      dimensions: [X_AXIS_DATA_KEY, WATERFALL_TOTAL_KEY],
+      encode: {
+        y: WATERFALL_TOTAL_KEY,
+        x: X_AXIS_DATA_KEY,
+      },
+      itemStyle: {
+        color: settings["waterfall.total_color"],
+      },
+      animationDuration: 0,
+    });
+  }
 
-export function getWaterfallOption(
-  chartModel: WaterfallChartModel,
+  return series;
+};
+
+export const getWaterfallChartOption = (
+  chartModel: BaseCartesianChartModel,
   timelineEventsModel: TimelineEventsModel | null,
   selectedTimelineEventsIds: TimelineEventId[],
   settings: ComputedVisualizationSettings,
   chartWidth: number,
+  chartHeight: number,
   renderingContext: RenderingContext,
-): EChartsOption {
-  const baseOption = getCartesianChartOption(
+): EChartsOption => {
+  const hasTimelineEvents = timelineEventsModel != null;
+  const chartMeasurements = getChartMeasurements(
     chartModel,
-    timelineEventsModel,
-    selectedTimelineEventsIds,
     settings,
+    hasTimelineEvents,
     chartWidth,
+    chartHeight,
+    renderingContext,
+  );
+  const timelineEventsSeries = hasTimelineEvents
+    ? getTimelineEventsSeries(
+        timelineEventsModel,
+        selectedTimelineEventsIds,
+        renderingContext,
+      )
+    : null;
+
+  const dataSeriesOptions = buildEChartsWaterfallSeries(
+    chartModel,
+    settings,
+    chartMeasurements,
     renderingContext,
   );
 
-  const dataset: DatasetOption = {
-    source: chartModel.waterfallDataset,
-    dimensions: Object.values(DATASET_DIMENSIONS),
-  };
+  const seriesOption = [dataSeriesOptions, timelineEventsSeries].flatMap(
+    option => option ?? [],
+  );
+
+  const echartsDataset = [{ source: chartModel.transformedDataset }];
 
   return {
-    ...baseOption,
-    dataset,
-    ...getAxes(
-      settings,
+    // TODO: extract common options
```

Comment:
Should we do this now or in another PR?

If it's another PR please file an issue so it's not forgotten